src/livekit/OpenIDLoader.tsx

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2023 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {
+  ReactNode,
+  createContext,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import {
+  OpenIDClientParts,
+  SFUConfig,
+  getSFUConfigWithOpenID,
+} from "./openIDSFU";
+import { ErrorView, LoadingView } from "../FullScreenView";
+
+interface Props {
+  client: OpenIDClientParts;
+  livekitServiceURL: string;
+  roomName: string;
+  children: ReactNode;
+}
+
+const SFUConfigContext = createContext<SFUConfig>(undefined);
+
+export const useSFUConfig = () => useContext(SFUConfigContext);
+
+export function OpenIDLoader({
+  client,
+  livekitServiceURL,
+  roomName,
+  children,
+}: Props) {
+  const [sfuConfig, setSFUConfig] = useState<SFUConfig>();
+  const [error, setError] = useState<Error>();
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const result = await getSFUConfigWithOpenID(
+          client,
+          livekitServiceURL,
+          roomName
+        );
+        setSFUConfig(result);
+      } catch (e) {
+        logger.error("Failed to fetch SFU config: ", e);
+        setError(new Error("Failed to fetch SFU config"));
+      }
+    })();
+  }, [client, livekitServiceURL, roomName]);
+
+  if (error) {
+    return <ErrorView error={error} />;
+  } else if (sfuConfig) {
+    return (
+      <SFUConfigContext.Provider value={sfuConfig}>
+        {children}
+      </SFUConfigContext.Provider>
+    );
+  } else {
+    return <LoadingView />;
+  }
+}
```

Comment:
Cool! A small hint: it seems like we have 3 distinct states here: loading that then resolves into **either** loaded **or** failed (it's not possible to have both `error` and `sfuConfig` set). Also, the current state won't really compile with strict lints since the actual type of each state is `SFUConfig | undefined` and `Error | undefined` respectively.

It seems to be a good idea to model the state explicitly (stronger types, no possibility to have mutually exclusive states) like this:

```suggestion
  const [state, setState] = useState<
    SFUConfigLoading | SFUConfigLoaded | SFUConfigFailed
  >({ kind: "loading" });

  useEffect(() => {
    (async () => {
      try {
        const result = await getSFUConfigWithOpenID(
          client,
          livekitServiceURL,
          roomName
        );
        setState({ kind: "loaded", sfuConfig: result });
      } catch (e) {
        logger.error("Failed to fetch SFU config: ", e);
        setState({ kind: "failed", error: e });
      }
    })();
  }, [client, livekitServiceURL, roomName]);

  switch (state.kind) {
    case "loading":
      return <LoadingView />;
    case "failed":
      return <ErrorView error={state.error} />;
    case "loaded":
      return (
        <SFUConfigContext.Provider value={state.sfuConfig}>
          {children}
        </SFUConfigContext.Provider>
      );
  }
}

type SFUConfigLoading = {
  kind: "loading";
};

type SFUConfigLoaded = {
  kind: "loaded";
  sfuConfig: SFUConfig;
};

type SFUConfigFailed = {
  kind: "failed";
  error: Error;
};
```

It prevents all kinds of problems (especially if the code gets larger, and has a more complicated state) and is also a bit more elegant 🙂

Reply:
oh, so have them as a union type: yeah, that works.